File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-java-Mothyque/backend/Server/src/main/java/org/example/networking/ServicesImpl.java
package org.example.networking;

import org.example.domain.*;
import org.example.service.ClientService;
import org.example.service.MatchService;
import org.example.service.TicketService;
import org.example.service.UserService;
import org.example.utils.Observer;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class ServicesImpl implements org.example.networking.Services
{
    private UserService userService;
    private MatchService matchService;
    private ClientService clientService;
    private TicketService ticketService;
    private Map<Integer, Observer> loggedClients;
    private Set<Integer> activeUsersIds;

    public ServicesImpl(UserService userService, MatchService matchService, ClientService clientService, TicketService ticketService) 
    {
        this.userService = userService;
        this.matchService = matchService;
        this.clientService = clientService;
        this.ticketService = ticketService;
        loggedClients = new ConcurrentHashMap<>();
        activeUsersIds = ConcurrentHashMap.newKeySet();
    }

    @Override
    public synchronized void login(User user, Observer client) throws Exception
    {
        User userFound = userService.findByUsername(user.getUsername());
        if (userFound != null)
        {
            if(activeUsersIds.contains(userFound.getId()))
            {
                throw new Exception("User already logged in.");
            }
            else
            {
                User loggedUser = userService.authenticate(user.getUsername(), user.getPassword());
                if (loggedUser != null)
                {
                    activeUsersIds.add(loggedUser.getId());
                    if (client != null)
                    {
                        loggedClients.put(loggedUser.getId(), client);
                    }
                }
                else
                {
                    throw new Exception("Invalid username or password.");
                }
            }

        }
    }

    @Override
    public synchronized void logout(User user, Observer client) throws Exception
    {
        User userFound = userService.findByUsername(user.getUsername());
        if (userFound == null) return;

        activeUsersIds.remove(userFound.getId());
        loggedClients.remove(userFound.getId());
    }

    @Override
    public Match[] getAllMatches() throws Exception
    {
        try
        {
            Iterable<Match> matches = matchService.findAll();
            List<Match> matchList = new ArrayList<>();
            matches.forEach(matchList::add);
            return matchList.toArray(new Match[0]);
        }
        catch (Exception e)
        {
            throw new Exception("Error retrieving matches: " + e.getMessage());
        }
    }

    @Override
    public synchronized void buyTickets(Match match, String clientName, String clientAddress, int numberOfTickets) throws Exception
    {
        Match selectedMatch = matchService.findOne(match.getId()).orElseThrow(() -> new Exception("Match not found."));
        if (selectedMatch.getAvailableSeats() < numberOfTickets)
        {
            throw new Exception("Not enough available seats for the selected match.");
        }
        Client client = clientService.findByNameAndAddress(clientName, clientAddress);
        if (client == null)
        {
            client = new Client(clientName, clientAddress);
            clientService.add(client);
            client = clientService.findByNameAndAddress(clientName, clientAddress);
        }

        int currentAvailable = selectedMatch.getAvailableSeats();
        for (int i = 0; i < numberOfTickets; i++)
        {
            int seatNumber = selectedMatch.getTotalSeats() - currentAvailable + i + 1;
            Ticket newTicket = new Ticket(client.getId(), selectedMatch.getId(), "Seat " + seatNumber);
            ticketService.add(newTicket);
        }

        selectedMatch.setAvailableSeats(currentAvailable - numberOfTickets);
        matchService.update(selectedMatch);

        notifyAllClients();
    }

    @Override
    public List<ClientTicketDTO> filterTickets(String name, String address) throws Exception
    {
        try
        {
            return ticketService.getTicketByNameAndAddress(name, address);
        }
        catch (Exception e)
        {
            throw new Exception("Error filtering tickets: " + e.getMessage());
        }
    }

    @Override
    public synchronized void updateTickets(Match match, Client client, int oldQuantity, int newQuantity) throws Exception
    {
        Match matchDb = matchService.findOne(match.getId()).orElseThrow(() -> new Exception("Match not found."));
        int diff = newQuantity - oldQuantity;

        if (diff > 0 && matchDb.getAvailableSeats() < diff)
        {
            throw new Exception("Not enough available seats for the match.");
        }

        ticketService.updateClientTickets(client, matchDb, diff);
        notifyAllClients();
    }

    @Override
    public synchronized void addObserver(Integer userId, Observer client)
    {
        loggedClients.put(userId, client);
    }

    private void notifyAllClients()
    {
        for (Map.Entry<Integer, Observer> entry : loggedClients.entrySet())
        {
            try
            {
                entry.getValue().update();
            }
            catch (Exception e)
            {
                System.err.println("Error notifying client: " + e.getMessage());
            }
        }
    }
}

File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-csharp-Mothyque/ProiectMPP/ProiectMPP.Server/Networking/ServicesImpl.cs
using ProiectMPP.Common.Domain;
using ProiectMPP.Common.dto;
using ProiectMPP.Common.Networking;
using ProiectMPP.Common.Utils;
using ProiectMPP.Server.Repository;
using ProiectMPP.Server.Service;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace ProiectMPP.Server
{
    public class ServicesImpl : IServices
    {
        private UserService userService;
        private MatchService matchService;
        private ClientService clientService;
        private TicketService ticketService;
        private IDictionary<int, IObserver> loggedClients;

        public ServicesImpl(UserService userService, MatchService matchService,
                            ClientService clientService, TicketService ticketService)
        {
            this.userService = userService;
            this.matchService = matchService;
            this.clientService = clientService;
            this.ticketService = ticketService;
            loggedClients = new ConcurrentDictionary<int, IObserver>();
        }

        public void Login(User user, IObserver client)
        {

            lock (this)
            {
                if (userService == null)
                {
                    throw new Exception("Server internal error: userService null.");
                }

                User userFound = userService.FindByUsername(user.Username);

                if (userFound != null)
                {
                    if (loggedClients.ContainsKey(userFound.Id))
                    {
                        throw new Exception("User already logged in.");
                    }

                    User loggedUser = userService.Authenticate(user.Username, user.Password);

                    if (loggedUser != null)
                    {
                        loggedClients[loggedUser.Id] = client;
                    }
                    else
                    {
                        throw new Exception("Invalid username or password.");
                    }
                }
                else
                {
                    throw new Exception("User not found.");
                }
            }
        }

        public void Logout(User user, IObserver client)
        {
            lock (this)
            {
                if (SharedConnectionManager.HasActiveTransaction())
                {
                    try
                    {
                        SharedConnectionManager.RollbackTransaction();
                    }
                    catch (Exception ex)
                    {
                    }
                }

                User userFound = null;
                try
                {
                    userFound = userService.FindByUsername(user.Username);
                }
                catch (Exception ex)
                {
                }

                if (userFound != null && loggedClients.Remove(userFound.Id))
                {
                    return;
                }

                throw new Exception("User not logged in.");
            }
        }

        public Match[] GetAllMatches()
        {
            try
            {
                IEnumerable<Match> matches = matchService.FindAll();
                return matches.ToArray();
            }
            catch (Exception e)
            {
                throw new Exception("Error retrieving matches: " + e.Message);
            }
        }

        public void BuyTickets(Match match, string clientName, string clientAddress, int numberOfTickets)
        {
            lock (this)
            {
                try
                {
                    ticketService.BeginTransaction();

                    Match selectedMatch = matchService.FindOne(match.Id) ??
                                         throw new Exception("Match not found.");

                    if (selectedMatch.AvailableSeats < numberOfTickets)
                    {
                        throw new Exception("Not enough available seats for the selected match.");
                    }

                    Client client = clientService.FindByNameAndAddress(clientName, clientAddress);
                    if (client == null)
                    {
                        client = new Client(clientName, clientAddress);
                        clientService.Add(client);
                        client = clientService.FindByNameAndAddress(clientName, clientAddress);
                    }

                    int currentAvailable = selectedMatch.AvailableSeats;
                    for (int i = 0; i < numberOfTickets; i++)
                    {
                        int seatNumber = selectedMatch.TotalSeats - currentAvailable + i + 1;
                        Ticket newTicket = new Ticket(client.Id, selectedMatch.Id, "Seat " + seatNumber);
                        ticketService.Add(newTicket);
                    }

                    selectedMatch.AvailableSeats = currentAvailable - numberOfTickets;
                    matchService.Update(selectedMatch);

                    ticketService.CommitTransaction();
                }
                catch (Exception ex)
                {
                    ticketService.RollbackTransaction();
                    throw;
                }
            }
            NotifyAllClients();
        }

        public IList<ClientTicketDTO> FilterTickets(string name, string address)
        {
            try
            {
                return ticketService.GetTicketByNameAndAddress(name, address).ToList();
            }
            catch (Exception e)
            {
                throw new Exception("Error filtering tickets: " + e.Message);
            }
        }

        public void UpdateTickets(Match match, Client client, int oldQuantity, int newQuantity)
        {
            lock (this)
            {
                try
                {
                    ticketService.BeginTransaction();

                    Match matchDb = matchService.FindOne(match.Id);
                    if (matchDb == null)
                    {
                        throw new Exception("Match not found.");
                    }

                    int diff = newQuantity - oldQuantity;

                    if (diff > 0 && matchDb.AvailableSeats < diff)
                    {
                        throw new Exception("Not enough available seats for the match.");
                    }

                    try
                    {
                        ticketService.UpdateClientTickets(client, matchDb, diff);
                    }
                    catch (Exception ex)
                    {
                        throw;
                    }

                    ticketService.CommitTransaction();
                }
                catch (Exception ex)
                {
                    ticketService.RollbackTransaction();
                    throw;
                }
            }
            NotifyAllClients();
        }

        private void NotifyAllClients()
        {
            foreach (var client in loggedClients.Values)
            {
                try
                {
                    client.Update();
                }
                catch (Exception e)
                {
                    Console.Error.WriteLine("Error notifying client: " + e.Message);
                }
            }
        }
    }
}